README.rst
=======================================================
Python SDK for IoT Data Exchange & Analytics Middleware
=======================================================

Key Features
============

- Supports IoT entities to register, publish, subscribe, access historical data.
- Supports asynchronous access of data from the entity.

Installation
============
    ``$ python3 -m pip install ideam``

**NOTE**: Python2 is not supported.

Getting started
===============
Using this FORM_, you can register yourself with the middleware.

After registering yourself as a provider of the devices/applications with the middleware, you will receive a ``OWNER API KEY``.


Register
--------
To **register** a new entity with the (IDEAM) Middleware, you need a ``UNIQUE device name`` and the ``OWNER API KEY``:

.. code-block:: python

  >>> from ideam.entity import Entity
  >>> deviceHandler = Entity("new-device-name", "OWNER-API-KEY")
  >>> deviceHandler.register()
  {'APIKey': 'DEVICE-API-KEY', 'Subscription Queue Name': 'new-device-name', 'ResourceID': 'new-device-name', 'Registration': 'success'}
  >>> deviceHandler.set_entity_api_key("DEVICE-API-KEY")
  DEVICE-API-KEY

**NOTE**: Store the ``DEVICE-API-KEY`` obtained in the response, in a safe place. This ``DEVICE-API-KEY`` will be required for doing operations like publish, subscribe and accessing historical data.

For an Already Registered Device
--------------------------------
Here, we can skip the registration of the entity with the middleware.

.. code-block:: python

  >>> from ideam.entity import Entity
  >>> deviceHandler = Entity("device-name", "OWNER-API-KEY")
  >>> deviceHandler.set_entity_api_key("DEVICE-API-KEY")


Publish
-------
To **publish** to the (IDEAM) Middleware as an IoT device/entity:



 Args:
            *data*    (string): contents to be published by this entity.

.. code-block:: python

  >>> from ideam.entity import Entity
  >>> deviceHandler = Entity("publishing-device-name", "OWNER-API-KEY")
  >>> deviceHandler.set_entity_api_key("PUBLISHING-DEVICE-API-KEY")
  >>> deviceHandler.publish("demo data")
  {'status': 'success', 'response': 'publish message ok'}


Subscribe
---------

To **listen (subscribe)** to the data from a device/entity registered with the (IDEAM) Middleware, do the following steps:

 **Subscribe**. This method will start listening for the data from middleware. This method will automatically bind or listen to any devices provided in the argument list. When a new data from the device arrives, it is stored in an internal variable called ``subscribe_data``.  The ``subscribe_data`` is a dictionary with ``data`` and ``timestamp`` field. ``timestamp`` denotes the epoch time (in milliseconds) at which data arrived.

 Args:
           *devices_to_bind*  (list of strings): an array of devices to listen to.
                                                 Ex: subscribe(["test100", "testDemo"])

.. code-block:: python

  >>> from ideam.entity import Entity
  >>> applicationHandler = Entity("listening-device-name", "OWNER-API-KEY")
  >>> applicationHandler.set_entity_api_key("LISTENING-DEVICE-API-KEY")
  >>>
  >>> applicationHandler.subscribe(["publishing-device-name", "publishing-device-2"])
  >>> applicationHandler.subscribe_data
  {'data': 'demo data', 'timestamp': '1513526954674'}


Unbind
------
To unbind any entity that is already bound to, use the **unbind** method:

  Args:
      *devices_to_unbind* (list of strings): an array of devices that are to be unbound ( stop listening). Ex. unbind(["test10","testDemo105"])

.. code-block:: python

  >>> from ideam.entity import Entity
  >>> applicationHandler = Entity("listening-device-name", "OWNER-API-KEY")
  >>> applicationHandler.set_entity_api_key("LISTENING-DEVICE-API-KEY")
  >>> applicationHandler.unbind(["publishing-device-name"])
  {'status': 'success', 'response': 'unbind queue ok'}

Access historical data
----------------------
The db function allows an entity to access the historical data.

   Args:
        entity        (string): Name of the device to listen
        query_filters (string): Elastic search response format string. Ex. query_filters="pretty=true&size=10"

.. code-block:: python

  >>> from ideam.entity import Entity
  >>> applicationHandler = Entity("listening-device-name", "OWNER-API-KEY")
  >>> applicationHandler.set_entity_api_key("LISTENING-DEVICE-API-KEY")
  >>> applicationHandler.db("rbccpsEnergy.EM_D0025860")
  '{"took":5,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":92292,"max_score":1.0487294,"hits":[{"_index":"sensor_data","_type":"logs","_id":"AV6AVeOG7sVBkWsIECvP","_score":1.0487294,"_source":{"@timestamp":"2017-09-14T12:21:06.047Z","data":"{\\"YPhaseReactivePower\\": 2407.9000949859619, \\"BPhaseVoltage\\": 239.428466796875, \\"YPhaseApparentPower\\": 3263.2999420166016, \\"YPhaseActivePower\\": 2202.8000354766846, \\"RPhasePowerFactor\\": 0.78799998760223389, \\"BPhaseActivePower\\": 2222.1999168395996, \\"EnergyReactive\\": 18639.000782012939, \\"BPhaseCurrent\\": 14.46090030670166, \\"RPhaseApparentPower\\": 5156.0001373291016, \\"RPhaseReactivePower\\": 3173.30002784729, \\"YPhasePowerFactor\\": 0.67400002479553223, \\"RPhaseVoltage\\": 234.58619689941406, \\"BPhaseReactivePower\\": 2654.9999713897705, \\"BPhasePowerFactor\\": 0.64099997282028198, \\"RPhaseActivePower\\": 4066.8997764587402, \\"YPhaseCurrent\\": 13.757100105285645, \\"YPhaseVoltage\\": 237.21040344238281, \\"RPhaseCurrent\\": 21.979299545288086, \\"BPhaseApparentPower\\": 3462.3000621795654, \\"dataSamplingInstant\\": 1505138556.0, \\"EnergyActive\\": 20038.0}","@version":"1","routing-key":"rbccpsEnergy.EM_D0025860","key":"rbccpsEnergy.EM_D0025860"}},{"_index":"sensor_data","_type":"logs","_id":"AV6AQ3-E7sVBkWsIECuo","_score":1.0487294,"_source":{"@timestamp":"2017-09-14T12:01:00.796Z","data":"{\\"YPhaseReactivePower\\": 2367.5999641418457, \\"BPhaseVoltage\\": 238.37925720214844, \\"YPhaseApparentPower\\": 3248.5001087188721, \\"YPhaseActivePower\\": 2224.600076675415, \\"RPhasePowerFactor\\": 0.79400002956390381, \\"BPhaseActivePower\\": 2253.4999847412109, \\"EnergyReactive\\": 18635.600391387939, \\"BPhaseCurrent\\": 14.405300140380859, \\"RPhaseApparentPower\\": 5144.4997787475586, \\"RPhaseReactivePower\\": 3123.1000423431396, \\"YPhasePowerFactor\\": 0.68400001525878906, \\"RPhaseVoltage\\": 233.84330749511719, \\"BPhaseReactivePower\\": 2590.8999443054199, \\"BPhasePowerFactor\\": 0.65600001811981201, \\"RPhaseActivePower\\": 4091.1998748779297, \\"YPhaseCurrent\\": 13.756699562072754, \\"YPhaseVoltage\\": 236.14106750488281, \\"RPhaseCurrent\\": 22.0, \\"BPhaseApparentPower\\": 3433.9001178741455, \\"dataSamplingInstant\\": 1505137324.0, \\"EnergyActive\\": 20034.201171875}","@version":"1","routing-key":"rbccpsEnergy.EM_D0025860","key":"rbccpsEnergy.EM_D0025860"}},{"_index":"sensor_data","_type":"logs","_id":"AV6AVBB27sVBkWsIECvK","_score":1.0487294,"_source":{"@timestamp":"2017-09-14T12:19:06.479Z","data":"{\\"YPhaseReactivePower\\": 0.0, \\"BPhaseVoltage\\": 0.0, \\"YPhaseApparentPower\\": 0.0, \\"YPhaseActivePower\\": 0.0, \\"RPhasePowerFactor\\": 0.0, \\"BPhaseActivePower\\": 0.0, \\"EnergyReactive\\": 0.0, \\"BPhaseCurrent\\": 0.0, \\"RPhaseApparentPower\\": 0.0, \\"RPhaseReactivePower\\": 0.0, \\"YPhasePowerFactor\\": 0.0, \\"RPhaseVoltage\\": 0.0, \\"BPhaseReactivePower\\": 0.0, \\"BPhasePowerFactor\\": 0.0, \\"RPhaseActivePower\\": 0.0, \\"YPhaseCurrent\\": 0.0, \\"YPhaseVoltage\\": 0.0, \\"RPhaseCurrent\\": 0.0, \\"BPhaseApparentPower\\": 0.0, \\"dataSamplingInstant\\": 1505138437.0, \\"EnergyActive\\": 0.0}","@version":"1","routing-key":"rbccpsEnergy.EM_D0025860","key":"rbccpsEnergy.EM_D0025860"}}]}}'


.. _FORM: https://docs.google.com/forms/d/e/1FAIpQLSc-L_kMayQjpXsIZ5BU_UCBFI_v6dNPrBcmQIHp0J3kBkfyFQ/viewform?c=0&w=1


CHANGES.rst
=========
Changelog
=========

0.0.2 (2017-12-19)
==================

- Fixed missing readme.rst in pypi distribution

0.0.1 (2017-12-19)
==================

- Initial Commit with ISC License

ideam/ideam.py
import contextlib
import requests
import warnings
import json
import asyncio
import aiohttp
from time import time
import threading
requests.packages.urllib3.disable_warnings()
from sys import platform


class Entity(object):
    """ Entity object represents any of the IoT entities registered with the RBCCPS Smart City middleware.
    It can do publish, subscribe, bind and unbind operations. Details of these operations are specified in
    the https://rbccps.org/smartcity .

    """
    def __init__(self, entity_id, entity_api_key):
        self.entity_id = entity_id
        self.owner_api_key = entity_api_key
        self.entity_api_key = ""
        self.base_url = "https://smartcity.rbccps.org/"
        self.subscribe_data = {}
        self.event_loop = asyncio.get_event_loop()

    def __del__(self):
        self.stop_subscribe()

    def set_base_url(self, value):
        self.base_url = value
        return self.base_url

    def set_entity_api_key(self, value):
        self.entity_api_key = value
        return self.entity_api_key

    def entity_api_key(self):
        return self.entity_api_key

    def subscribe_data(self):
        return self.subscribe_data

    def register(self):
        """ Registers a new device in the format of entity_id. This device has permissions for services like
        subscribe, publish, historicData.

        Site: http://rbccps.org/smartcity
        """
        register_url = self.base_url + "api/0.1.0/register"
        register_headers = {
            "apikey": str(self.owner_api_key),
            "resourceID": str(self.entity_id),
            "serviceType": "publish,subscribe,historicData"
        }
        with self.no_ssl_verification():
            r = requests.get(register_url, {}, headers=register_headers)
        response = r.content.decode("utf-8")
        if "APIKey" in str(r.content.decode("utf-8")):
            response = json.loads(response[:-331] + "}")  # Temporary fix to a middleware bug, should be removed in future
            response["Registration"] = "success"
        else:
            response = json.loads(response)
            response["Registration"] = "failure"
        return response

    @contextlib.contextmanager
    def no_ssl_verification(self):
        """ Requests module fails due to lets encrypt ssl encryption. """
        try:
            from functools import partialmethod
        except ImportError:
            # Python 2 fallback: https://gist.github.com/carymrobbins/8940382
            from functools import partial

            class partialmethod(partial):
                def __get__(self, instance, owner):
                    if instance is None:
                        return self

                    return partial(self.func, instance, *(self.args or ()), **(self.keywords or {}))

        old_request = requests.Session.request
        requests.Session.request = partialmethod(old_request, verify=False)
        warnings.filterwarnings('ignore', 'Unverified HTTPS request')
        yield
        warnings.resetwarnings()
        requests.Session.request = old_request

    def publish(self, data):
        """ This function allows an entity to publish data to the middleware.

        Args:
            data    (string): contents to be published by this entity.
        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}
        publish_url = self.base_url + "api/0.1.0/publish"
        publish_headers = {"apikey": self.entity_api_key}
        publish_data = {
            "exchange": "amq.topic",
            "key": str(self.entity_id),
            "body": str(data)
        }
        with self.no_ssl_verification():
            r = requests.post(publish_url, json.dumps(publish_data), headers=publish_headers)
        response = dict()
        if "No API key" in str(r.content.decode("utf-8")):
            response["status"] = "failure"
            r = json.loads(r.content.decode("utf-8"))['message']
        elif 'publish message ok' in str(r.content.decode("utf-8")):
            response["status"] = "success"
            r = r.content.decode("utf-8")
        else:
            response["status"] = "failure"
            r = r.content.decode("utf-8")
        response["response"] = str(r)
        return response

    def db(self, entity, query_filters="size=10"):
        """"""
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}

        historic_url = self.base_url + "api/0.1.0/historicData?" + query_filters
        historic_headers = {
            "apikey": self.entity_api_key,
            "Content-Type": "application/json"
        }

        historic_query_data = json.dumps({
            "query": {
                "match": {
                    "key": entity
                }
            }
        })

        with self.no_ssl_verification():
            r = requests.get(historic_url, data=historic_query_data, headers=historic_headers)
        response = dict()
        if "No API key" in str(r.content.decode("utf-8")):
            response["status"] = "failure"
            r = json.loads(r.content.decode("utf-8"))['message']
        else:
            r = r.content.decode("utf-8")
            response = r
        return response

    def bind(self, devices_to_bind):
        """ This function allows an entity to list the devices to subscribe for data. This function must be called
        at least once, before doing a subscribe. Subscribe function will listen to devices that are bound here.

        Args:
            devices_to_bind  (list): a array of devices to listen to.
                                     Example bind(["test100","testDemo"])

        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}
        url = self.base_url + "api/0.1.0/subscribe/bind"
        headers = {"apikey": self.entity_api_key}
        data = {
            "exchange": "amq.topic",
            "keys": devices_to_bind,
            "queue": self.entity_id
        }

        with self.no_ssl_verification():
            r = requests.post(url, json=data, headers=headers)
        response = dict()
        if "No API key" in str(r.content.decode("utf-8")):
            response["status"] = "failure"
            r = json.loads(r.content.decode("utf-8"))['message']
        elif 'bind queue ok' in str(r.content.decode("utf-8")):
            response["status"] = "success"
            r = r.content.decode("utf-8")
        else:
            response["status"] = "failure"
            r = r.content.decode("utf-8")
        response["response"] = str(r)
        return response

    def unbind(self, devices_to_unbind):
        """ This function allows an entity to unbound devices that are already bound.

        Args:
            devices_to_unbind (list): a array of devices that are to be unbound ( stop listening)
                                     Example unbind(["test10","testDemo105"])

        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}
        url = self.base_url + "api/0.1.0/subscribe/unbind"
        headers = {"apikey": self.entity_api_key}
        data = {
            "exchange": "amq.topic",
            "keys": devices_to_unbind,
            "queue": self.entity_id
        }

        with self.no_ssl_verification():
            r = requests.delete(url, json=data, headers=headers)
            print(r)
        response = dict()
        if "No API key" in str(r.content.decode("utf-8")):
            response["status"] = "failure"
            r = json.loads(r.content.decode("utf-8"))['message']
        elif 'unbind' in str(r.content.decode("utf-8")):
            response["status"] = "success"
            r = r.content.decode("utf-8")
        else:
            response["status"] = "failure"
            r = r.content.decode("utf-8")
        response["response"] = str(r)
        return response

    def subscribe(self, devices_to_bind=[]):
        """ This function allows an entity to subscribe for data from the devices specified in the bind operation. It
        creates a thread with an event loop to manager the tasks created in start_subscribe_worker.
        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}
        self.bind(devices_to_bind)
        loop = asyncio.new_event_loop()
        t1 = threading.Thread(target=self.start_subscribe_worker, args=(loop,))
        t1.daemon = True
        t1.start()


    def start_subscribe_worker(self, loop):
        """Switch to new event loop as a thread and run until complete"""
        url = self.base_url + "api/0.1.0/subscribe"
        print("0")
        task = loop.create_task(self.asynchronously_get_data(url + "?name={0}".format(self.entity_id)))
        asyncio.set_event_loop(loop)
        loop.run_until_complete(task)
        self.event_loop = loop


    async def asynchronously_get_data(self, url):
        """ Asynchronously get data from Chunked transfer encoding of https://smartcity.rbccps.org/api/0.1.0/subscribe.
        (Only this function requires Python 3. Rest of the functions can be run in python2.
        Args:
             url (string): url to subscribe to
        """
        print(self.entity_api_key)
        headers = {"apikey": self.entity_api_key}
        print("1")
        print(url)
        try:
            print("2")
            async with aiohttp.ClientSession(connector=aiohttp.TCPConnector(verify_ssl=False)) as session:
                print("3")
                async with session.get(url, headers=headers, timeout=300) as response:
                    print("4")
                    while True:  # loop over for each chunk of data
                        chunk = await response.content.readchunk()
                        if not chunk:
                            break
                        if platform == "linux" or platform == "linux2": # In linux systems, readchunk() returns a tuple
                            chunk = chunk[0]
                        resp = dict()
                        resp["data"] = chunk.decode()
                        current_milli_time = lambda: int(round(time() * 1000))
                        resp["timestamp"] = str(current_milli_time())
                        self.subscribe_data = resp
        except Exception as e:
            print("\n*********    Oops: " + url + " " + str(type(e)) + str(e) + "     *********\n")
        print('\n*********    Closing TCP: {}     *********\n'.format(url))

    def stop_subscribe(self):
        """ This function is used to stop the event loop created when subscribe is called. But this function doesn't
        stop the thread and should be avoided until its completely developed.

        """
        asyncio.gather(*asyncio.Task.all_tasks()).cancel()
        self.event_loop.stop()
        self.event_loop.close()







ideam/entity.py
import contextlib
import requests
import warnings
import json
import asyncio
import aiohttp
from time import time
import threading
requests.packages.urllib3.disable_warnings()
from sys import platform


class Entity(object):
    """ The entity object represents any of the IoT entities registered with the RBCCPS IoT Data Exchange & Analytics
    Middleware (IDEAM). It can do publish, subscribe, historical data, bind and unbind operations.
    Details of these operations are specified in the https://rbccps.org/smartcity.

    """
    def __init__(self, entity_id, entity_api_key):
        self.entity_id = entity_id
        self.owner_api_key = entity_api_key
        self.entity_api_key = ""
        self.base_url = "https://smartcity.rbccps.org/"
        self.subscribe_data = {}
        self.event_loop = asyncio.get_event_loop()

    def __del__(self):
        self.stop_subscribe()

    def set_base_url(self, value):
        self.base_url = value
        return self.base_url

    def set_entity_api_key(self, value):
        self.entity_api_key = value
        return self.entity_api_key

    def entity_api_key(self):
        return self.entity_api_key

    def subscribe_data(self):
        return self.subscribe_data

    def register(self):
        """ Registers a new device with the name entity_id. This device has permissions for services like subscribe,
        publish and access historical data.

        """
        register_url = self.base_url + "api/0.1.0/register"
        register_headers = {
            "apikey": str(self.owner_api_key),
            "resourceID": str(self.entity_id),
            "serviceType": "publish,subscribe,historicData"
        }
        with self.no_ssl_verification():
            r = requests.get(register_url, {}, headers=register_headers)
        response = r.content.decode("utf-8")
        if "APIKey" in str(r.content.decode("utf-8")):
            response = json.loads(response[:-331] + "}")  # Temporary fix to a middleware bug, should be removed in future
            response["Registration"] = "success"
        else:
            response = json.loads(response)
            response["Registration"] = "failure"
        return response

    @contextlib.contextmanager
    def no_ssl_verification(self):
        """ Requests module fails due to lets encrypt ssl encryption. Will be fixed in the future release."""
        try:
            from functools import partialmethod
        except ImportError:
            # Python 2 fallback: https://gist.github.com/carymrobbins/8940382
            from functools import partial

            class partialmethod(partial):
                def __get__(self, instance, owner):
                    if instance is None:
                        return self

                    return partial(self.func, instance, *(self.args or ()), **(self.keywords or {}))

        old_request = requests.Session.request
        requests.Session.request = partialmethod(old_request, verify=False)
        warnings.filterwarnings('ignore', 'Unverified HTTPS request')
        yield
        warnings.resetwarnings()
        requests.Session.request = old_request

    def publish(self, data):
        """ This function allows an entity to publish data to the middleware.

        Args:
            data    (string): contents to be published by this entity.
        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}
        publish_url = self.base_url + "api/0.1.0/publish"
        publish_headers = {"apikey": self.entity_api_key}
        publish_data = {
            "exchange": "amq.topic",
            "key": str(self.entity_id),
            "body": str(data)
        }
        with self.no_ssl_verification():
            r = requests.post(publish_url, json.dumps(publish_data), headers=publish_headers)
        response = dict()
        if "No API key" in str(r.content.decode("utf-8")):
            response["status"] = "failure"
            r = json.loads(r.content.decode("utf-8"))['message']
        elif 'publish message ok' in str(r.content.decode("utf-8")):
            response["status"] = "success"
            r = r.content.decode("utf-8")
        else:
            response["status"] = "failure"
            r = r.content.decode("utf-8")
        response["response"] = str(r)
        return response

    def db(self, entity, query_filters="size=10"):
        """ This function allows an entity to access the historic data.

        Args:
            entity        (string): Name of the device to listen to
            query_filters (string): Elastic search response format string
                                    example, "pretty=true&size=10"
        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}

        historic_url = self.base_url + "api/0.1.0/historicData?" + query_filters
        historic_headers = {
            "apikey": self.entity_api_key,
            "Content-Type": "application/json"
        }

        historic_query_data = json.dumps({
            "query": {
                "match": {
                    "key": entity
                }
            }
        })

        with self.no_ssl_verification():
            r = requests.get(historic_url, data=historic_query_data, headers=historic_headers)
        response = dict()
        if "No API key" in str(r.content.decode("utf-8")):
            response["status"] = "failure"
        else:
            r = r.content.decode("utf-8")
            response = r
        return response

    def bind(self, devices_to_bind):
        """ This function allows an entity to list the devices to subscribe for data. This function must be called
        at least once, before doing a subscribe. Subscribe function will listen to devices that are bound here.

        Args:
            devices_to_bind  (list): an array of devices to listen to.
                                     Example bind(["test100","testDemo"])

        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}
        url = self.base_url + "api/0.1.0/subscribe/bind"
        headers = {"apikey": self.entity_api_key}
        data = {
            "exchange": "amq.topic",
            "keys": devices_to_bind,
            "queue": self.entity_id
        }

        with self.no_ssl_verification():
            r = requests.post(url, json=data, headers=headers)
        response = dict()
        if "No API key" in str(r.content.decode("utf-8")):
            response["status"] = "failure"
            r = json.loads(r.content.decode("utf-8"))['message']
        elif 'bind queue ok' in str(r.content.decode("utf-8")):
            response["status"] = "success"
            r = r.content.decode("utf-8")
        else:
            response["status"] = "failure"
            r = r.content.decode("utf-8")
        response["response"] = str(r)
        return response

    def unbind(self, devices_to_unbind):
        """ This function allows an entity to unbound devices that are already bound.

        Args:
            devices_to_unbind (list): an array of devices that are to be unbound ( stop listening)
                                     Example unbind(["test10","testDemo105"])
        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}
        url = self.base_url + "api/0.1.0/subscribe/unbind"
        headers = {"apikey": self.entity_api_key}
        data = {
            "exchange": "amq.topic",
            "keys": devices_to_unbind,
            "queue": self.entity_id
        }

        with self.no_ssl_verification():
            r = requests.delete(url, json=data, headers=headers)
            print(r)
        response = dict()
        if "No API key" in str(r.content.decode("utf-8")):
            response["status"] = "failure"
            r = json.loads(r.content.decode("utf-8"))['message']
        elif 'unbind' in str(r.content.decode("utf-8")):
            response["status"] = "success"
            r = r.content.decode("utf-8")
        else:
            response["status"] = "failure"
            r = r.content.decode("utf-8")
        response["response"] = str(r)
        return response

    def subscribe(self, devices_to_bind=[]):
        """ This function allows an entity to subscribe for data from the devices specified in the bind operation. It
        creates a thread with an event loop to manager the tasks created in start_subscribe_worker.

        Args:
            devices_to_bind (list): an array of devices to listen to
        """
        if self.entity_api_key == "":
            return {'status': 'failure', 'response': 'No API key found in request'}
        self.bind(devices_to_bind)
        loop = asyncio.new_event_loop()
        t1 = threading.Thread(target=self.start_subscribe_worker, args=(loop,))
        t1.daemon = True
        t1.start()


    def start_subscribe_worker(self, loop):
        """ Switch to new event loop as a thread and run until complete. """
        url = self.base_url + "api/0.1.0/subscribe"
        task = loop.create_task(self.asynchronously_get_data(url + "?name={0}".format(self.entity_id)))
        asyncio.set_event_loop(loop)
        loop.run_until_complete(task)
        self.event_loop = loop


    async def asynchronously_get_data(self, url):
        """ Asynchronously get data from Chunked transfer encoding of https://smartcity.rbccps.org/api/0.1.0/subscribe.
        (Only this function requires Python 3. Rest of the functions can be run in python2.

        Args:
             url (string): url to subscribe
        """
        headers = {"apikey": self.entity_api_key}
        try:
            async with aiohttp.ClientSession(connector=aiohttp.TCPConnector(verify_ssl=False)) as session:
                async with session.get(url, headers=headers, timeout=3000) as response:
                    while True:  # loop over for each chunk of data
                        chunk = await response.content.readchunk()
                        if not chunk:
                            break
                        if platform == "linux" or platform == "linux2": # In linux systems, readchunk() returns a tuple
                            chunk = chunk[0]
                        resp = dict()
                        resp["data"] = chunk.decode()
                        current_milli_time = lambda: int(round(time() * 1000))
                        resp["timestamp"] = str(current_milli_time())
                        self.subscribe_data = resp
        except Exception as e:
            print("\n*********    Oops: " + url + " " + str(type(e)) + str(e) + "     *********\n")
        print('\n*********    Closing TCP: {}     *********\n'.format(url))

    def stop_subscribe(self):
        """ This function is used to stop the event loop created when subscribe is called. But this function doesn't
        stop the thread and should be avoided until its completely developed.

        """
        asyncio.gather(*asyncio.Task.all_tasks()).cancel()
        self.event_loop.stop()
        self.event_loop.close()
